src/middleware.ts

```diff
@@ -106,3 +107,84 @@ export const withWebhookSignatureVerified =
 
     return handler(request);
   };
+
+export interface DashboardTokenPayload extends JwtPayload {
+  app: string;
+}
+
+export const withJWTVerified =
+  (getAppId: (request: Request) => Promise<string | undefined>): Middleware =>
+  (handler) =>
+  async (request) => {
+    const { [SALEOR_DOMAIN_HEADER]: saleorDomain, "authorization-bearer": token } = request.headers;
+
+    if (token === undefined) {
+      return Response.BadRequest({
+        success: false,
+        message: "Missing token.",
+      });
+    }
+
+    let tokenClaims;
+    try {
+      tokenClaims = jwt.decode(token as string);
+    } catch (e) {
+      console.error(e);
+      return Response.BadRequest({
+        success: false,
+        message: "Invalid token.",
+      });
+    }
+
+    if (tokenClaims === null) {
+      return Response.BadRequest({
+        success: false,
+        message: "Invalid token.",
+      });
+    }
+
+    if ((tokenClaims as DashboardTokenPayload).iss !== saleorDomain) {
+      return Response.BadRequest({
+        success: false,
+        message: "Invalid token.",
+      });
+    }
+
+    let appId: string | undefined;
+    try {
+      appId = await getAppId(request);
+    } catch (error) {
+      console.error("Error during getting the app ID.");
+      console.error(error);
+      return Response.BadRequest({
+        success: false,
+        message: "Error during token invalidation - could not obtain the app ID.",
+      });
+    }
+
+    if (!appId || (tokenClaims as DashboardTokenPayload).app !== appId) {
```

Comment:
if `!appId` - how does it mean, that token is invalid? shouldn't that be considered a 500 Internal Server Error, if our backend asking for `appId` using stored `auth_token` cannot resolve that to something defined?

Reply:
You are right, such error would be much better